extension/content/components/recipes/form/arguments/MessagingExperiment.js

```diff
@@ -0,0 +1,214 @@
+
+      <ToggleField label="Prevent New Enrollment" name="isEnrollmentPaused">
+        Prevents new users from joining the experiment cohort.
+        <br />
+        Existing users will remain in the experiment.
+      </ToggleField>
+
+      <FormGroup>
+        <Row>
+          <Col xs={24}>
+            <Branches />
+          </Col>
+        </Row>
+      </FormGroup>
+    </>
+  );
+}
+
+function Branches() {
+  const data = useRecipeDetailsData();
+  const dispatch = useRecipeDetailsDispatch();
+  const { preferenceType } = data.arguments;
+
+  const handleClickAddBranch = () => {
+    const newBranch = { ratio: 1, slug: "", value: {}, groups: [] };
+    if (preferenceType === "boolean") {
+      newBranch.value = false;
+    }
+
+    dispatch({
+      type: ACTION_UPDATE_DATA,
+      data: {
+        ...data,
+        arguments: {
+          ...data.arguments,
+          branches: [...data.arguments.branches, newBranch],
+        },
+      },
+    });
+  };
+
+  let branchesList = (
+    <>
+      <HelpBlock>There are no branches.</HelpBlock>
+      <Divider />
+    </>
+  );
+  if (data.arguments.branches && data.arguments.branches.length) {
+    branchesList = data.arguments.branches.map((branch, index) => (
+      <Branch key={index} index={index} />
+    ));
+  }
+
+  return (
+    <FormGroup>
+      <ControlLabel>Branches</ControlLabel>
+      <Panel bordered>
+        {branchesList}
+        <IconButton
+          icon={<Icon icon="plus-circle" />}
+          onClick={handleClickAddBranch}
+        >
+          Add Branch
+        </IconButton>
+      </Panel>
+    </FormGroup>
+  );
+}
+
+function Branch({ index }) {
+  const data = useRecipeDetailsData();
+  const dispatch = useRecipeDetailsDispatch();
+  const branch = data.arguments.branches[index];
+  const { branches } = data.arguments;
+  const [branchesUpdated, setBranchesUpdated] = React.useState(Date.now());
+
+  const handleClickDelete = () => {
+    setBranchesUpdated(Date.now());
+    dispatch({
+      type: ACTION_UPDATE_DATA,
+      data: {
+        ...data,
+        arguments: {
+          ...data.arguments,
+          branches: branches.filter((b, i) => i !== index),
+        },
+      },
+    });
+  };
+
+  const handleChange = (name, cb = (v) => v) => {
+    return (value) => {
+      dispatch({
+        type: ACTION_UPDATE_DATA,
+        data: {
+          ...data,
+          arguments: {
+            ...data.arguments,
+            branches: branches.map((b, i) => {
+              if (i === index) {
+                return { ...b, [name]: cb(value) };
+              }
+
+              return b;
+            }),
+          },
+        },
+      });
+    };
+  };
+
+  const parseNumericInput = (value) => {
+    if (!value) {
+      return "";
+    }
+
+    return parseInt(value, 10);
+  };
+
+  let valueFieldKey = data.recipe && data.recipe.id ? data.recipe.id : "create";
+  valueFieldKey += `-${index}-${branchesUpdated}`;
```

Comment:
this was a tricky thing to sort out but makes sure that when you delete a branch the value json editor field updates correctly